core/src/apps/cardano/sign_tx/signer.py

```diff
@@ -0,0 +1,903 @@
+from ..helpers.credential import Credential, should_show_credentials
+from ..helpers.hash_builder_collection import HashBuilderDict, HashBuilderList
+from ..helpers.paths import (
+    CERTIFICATE_PATH_NAME,
+    CHANGE_OUTPUT_PATH_NAME,
+    CHANGE_OUTPUT_STAKING_PATH_NAME,
+    POOL_OWNER_STAKING_PATH_NAME,
+    SCHEMA_STAKING,
+)
+from ..helpers.utils import (
+    derive_public_key,
+    get_public_key_hash,
+    validate_network_info,
+    validate_stake_credential,
+)
+from ..seed import is_byron_path, is_minting_path, is_multisig_path, is_shelley_path
+
+if TYPE_CHECKING:
+    from typing import Any
+    from apps.common.paths import PathSchema
+
+    from .. import seed
+
+    CardanoTxResponseType = (
+        messages.CardanoTxItemAck | messages.CardanoTxWitnessResponse
+    )
+
+MINTING_POLICY_ID_LENGTH = 28
+MAX_ASSET_NAME_LENGTH = 32
+
+TX_BODY_KEY_INPUTS = const(0)
+TX_BODY_KEY_OUTPUTS = const(1)
+TX_BODY_KEY_FEE = const(2)
+TX_BODY_KEY_TTL = const(3)
+TX_BODY_KEY_CERTIFICATES = const(4)
+TX_BODY_KEY_WITHDRAWALS = const(5)
+TX_BODY_KEY_AUXILIARY_DATA = const(7)
+TX_BODY_KEY_VALIDITY_INTERVAL_START = const(8)
+TX_BODY_KEY_MINT = const(9)
+TX_BODY_KEY_SCRIPT_DATA_HASH = const(11)
+TX_BODY_KEY_COLLATERAL_INPUTS = const(13)
+TX_BODY_KEY_REQUIRED_SIGNERS = const(14)
+TX_BODY_KEY_NETWORK_ID = const(15)
+
+POOL_REGISTRATION_CERTIFICATE_ITEMS_COUNT = 10
+
+
+class Signer:
+    """
+    This class encapsulates the entire tx signing process. By default, most tx items are allowed
+    and shown to the user. For each signing mode, there is a subclass that overrides some methods,
+    usually to add more validation rules and hide some items.
+    Each tx item is processed in a _process_xyz() method which handles validation, user
+    confirmation and serialization of the tx item.
+    """
+
+    def __init__(
+        self,
+        ctx: wire.Context,
+        msg: messages.CardanoSignTxInit,
+        keychain: seed.Keychain,
+    ) -> None:
+        self.ctx = ctx
+        self.msg = msg
+        self.keychain = keychain
+
+        self.account_path_checker = AccountPathChecker()
+
+        # inputs, outputs and fee are mandatory, count the number of optional fields present
+        tx_dict_items_count = 3 + sum(
+            (
+                msg.ttl is not None,
+                msg.certificates_count > 0,
+                msg.withdrawals_count > 0,
+                msg.has_auxiliary_data,
+                msg.validity_interval_start is not None,
+                msg.minting_asset_groups_count > 0,
+                msg.include_network_id,
+                msg.script_data_hash is not None,
+                msg.collateral_inputs_count > 0,
+                msg.required_signers_count > 0,
+            )
+        )
+        self.tx_dict: HashBuilderDict[int, Any] = HashBuilderDict(
+            tx_dict_items_count, wire.ProcessError("Invalid tx signing request")
+        )
+
+    async def sign(self) -> None:
```

Comment:
If you mean the `sign` function itself, `_process_xxx` methods and a few helper functions at the bottom, then sure. I don't think it is a good idea to mark `_validate_xxx` and `_show_xxx` methods that happen to not be overriden anywhere - as I understand the `@final` decorator, it describes the intent, meaning "this should not be overriden", not the current state of the subclasses (also, [the relevant PEP](https://peps.python.org/pep-0591/) mentions no such use case), so it might be confusing.

Reply:
I added the decorator to the methods that I listed in the previous comment.

Reply:
It turns out that Trezor has their own fake `typing` module (because of build optimization) and this doesn't play out well with decorators from `typing` (also, no other app uses the `@final` decorator). Making this work would require modifications in the fake `typing` module and I guess it's not worth it.